Python/Product/Analysis/Interpreter/SharedDatabaseState.cs

```diff
@@ -177,34 +176,71 @@ internal SharedDatabaseState(Version languageVersion, string databaseDirectory,
             var obj = ObjectType;
             if (obj != null) {
                 assign(obj);
-            } else if (_objectTypeFixups != null) {
-                _objectTypeFixups.Add(assign);
-            } else {
-                throw new InvalidOperationException("Cannot find builtin type 'object'");
+                return;
             }
+
+            var fixups = _objectTypeFixups;
+            if (fixups != null) {
+                lock (fixups) {
+                    // Ensure it was not changed while waiting for the lock
+                    if (ReferenceEquals(_objectTypeFixups, fixups)) {
+                        fixups.Add(assign);
+                        return;
+                    }
+                }
+
+                // _objectTypeFixups changed, and it only changes to null, so we
+                // should call assign directly
+                obj = ObjectType;
+                if (obj != null) {
+                    assign(obj);
+                    return;
+                }
+            } 
+
+            throw new InvalidOperationException("Cannot find builtin type 'object'");
         }
 
-        private void RunObjectTypeFixups() {
-            if (_objectType == null) {
-                _objectType = BuiltinModule.GetAnyMember(GetBuiltinTypeName(BuiltinTypeId.Object)) as IPythonType;
-                if (_objectType != null) {
-                    var fixups = _objectTypeFixups;
-                    _objectTypeFixups = null;
-                    if (fixups != null) {
-                        foreach (var assign in fixups) {
-                            assign(_objectType);
-                        }
+        private IPythonType RunObjectTypeFixups() {
+            var objectType = _objectType;
+            if (objectType == null) {
+                var newObjectType = BuiltinModule.GetAnyMember(GetBuiltinTypeName(BuiltinTypeId.Object)) as IPythonType;
+
+                if (newObjectType == null) {
+                    // No type available, so don't do fixups
+                    return null;
+                }
+
+                // Either set _objectType to newObjectType, or replace our local
+                // objectType with whatever got there first
+                objectType = Interlocked.CompareExchange(ref _objectType, newObjectType, null) ?? newObjectType;
+            }
+
+            var fixups = _objectTypeFixups;
+            if (fixups != null) {
+                lock (fixups) {
+                    if (!ReferenceEquals(_objectTypeFixups, fixups)) {
+                        // _objectTypeFixups changed while we waited to lock
+                        return objectType;
                     }
+                    _objectTypeFixups = null;
+                }
+
+                foreach (var assign in fixups) {

```

Comment:
We've cleared the field inside the lock, and the adds check that the field has not been cleared before adding. Since the check/add is entirely within the lock, we know that nobody else has a reference to the list after we lock/clear. (At least, that's how I think it works. Very hard to reproduce this race...)